[MPM] Restart Application

andinoboerst opened the pull request:
**📝 Description**
Implementing the restart application for use with the Particle Mechanics Application.

**🆕 Changelog**
- Enabled the mpm_solver to identify and accept restart files as model imports
- Forced reassignment of the material points to the corresponding grid elements on restarts

andinoboerst committed:
Assigning all Material Points to Grid Elements

andinoboerst committed:
Modifying mpm_solver to enable restarts

philbucher commented on applications/ParticleMechanicsApplication/custom_utilities/mpm_search_element_utility.h:
```diff
@@ -429,8 +440,24 @@ namespace MPMSearchElementUtility
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(rMPMModelPart.Elements().size()); ++i) {
+                auto element_itr = (rMPMModelPart.ElementsBegin() + i);
+                #pragma omp critical
```
this serializes the entire loop, making it even slower than without OMP :)

I recommend to directly use `std::vector<typename Condition::Pointer> = rMPMModelPart.Conditions()`. AFAIK this is not possible (due to type mismatch), hence the easiest and fastest solution is probably the following:
~~~c++
missing_elements.resize(rMPMModelPart.NumberOFElements());
IndexPartition(rMPMModelPart.NumberOFElements()).for_each([](std::size_t i){
    missing_elements[i] = &*(rMPMModelPart.ElementsBegin() + i); // maybe add/remove *&?
});

andinoboerst replied:
True, good point.
I will try that solution and update, thank you! :)

andinoboerst replied:
Ok done :thumbsup:

andinoboerst committed:
Improving assignation of missing elements and conditions

andinoboerst committed:
Setting restarted back to false after first loop

andinoboerst committed:
Don't initialize the elements when restarting

andinoboerst committed:
Updated connection of grid and MP time step info

andinoboerst committed:
Deleting old commented code

andinoboerst committed:
Deleted trailing white spaces

andinoboerst committed:
Avoid reassigning gravity when restarting

andinoboerst committed:
Include restart tests

andinoboerst committed:
removed unused variable

andinoboerst committed:
Not reassigning buffer size on restart

andinoboerst committed:
Corrected element initialization on restart

andinoboerst committed:
Skip assigning initial velocities on restart

andinoboerst committed:
fixed if loop mistake

andinoboerst committed:
changed is_restarted flag in element initialization

andinoboerst committed:
update UP element

andinoboerst committed:
fixed error

andinoboerst committed:
changed restart check to comply with other element calls

philbucher commented on applications/ParticleMechanicsApplication/custom_constitutive/hyperelastic_3D_law.cpp:
```diff
@@ -153,10 +153,19 @@ void HyperElastic3DLaw::InitializeMaterial( const Properties& rMaterialPropertie
+    bool restarted = false;
+    
+    if (rShapeFunctionsValues.size() > 0){
+        if (rShapeFunctionsValues(0) != -888 ){
```
this is quite a hack :thinking: 
@AlejandroCornejo how do you handle this in the CLawsApp?
This function seems to be one of the few (if not the only one) that does not have access to the `ProcessInfo`

andinoboerst replied:
Yes 😅 i've been trying to figure out a way to solve this differently, for example by adding a "IS_RESTARTED" value to the Material Properties but I was running into issues when running it with multiple threads in OMP. So any other suggestions are welcome, I just used this as a temporary solution until I could come up with something different.

philbucher replied:
I did not look at it in detail, but I know that this would not be the first function in Elements/Conditions/Constitutive-Laws that is missing the `ProcessInfo

Lets wait for the response of @AlejandroCornejo
The proper solution would probably be to update the interface of this function. This involves the core and hence takes some time, but we have a workflow/knowledge how to do it backward compatible (e.g. see #2817)

philbucher replied:
> Material Properties but I was running into issues when running it with multiple threads in OMP

yes this is not allowed, and will lead to race conditions as you experienced :)

andinoboerst committed:
Corrected if statement

AlejandroCornejo commented on applications/ParticleMechanicsApplication/custom_constitutive/hencky_plastic_3D_law.cpp:
```diff
@@ -189,7 +189,7 @@ void HenckyElasticPlastic3DLaw::InitializeMaterial(const Properties& rProps,
 
     if (rShapeFunctionsValues.size() > 0){
-        if (rShapeFunctionsValues(0) != -888 ){
+        if (rShapeFunctionsValues(0) == -888 ){
```
I do not understand why you need to set anything related to the ShapeFunctions in the CL, can you develop a bit?

AlejandroCornejo replied:
@philbucher don't we have a `rProcessInfo[IS_RESTARTED]` capability?

rubenzorrilla replied:
Yes. Indeed, this is the purposed it is intended to be used for.

andinoboerst replied:
Yes however, I don't know how to access the ProcessInfo from within this function, as it is not passed in the function call and I could not find it being saved within the class (I might be wrong).

rubenzorrilla replied:
True. I think you'll need to do it through the properties...

rubenzorrilla replied:
Indeed, I can't understand why we are not passing the `ProcessInfo` in here. @KratosMultiphysics/technical-committee .

AlejandroCornejo replied:
Theoretically, the Initialize is done only once. If it is restarted it does not make sense to do it again (IMO). All the parameters to be initialized are serialized

andinoboerst replied:
I thought so too, but if I don't reinitialize the `mpMPMFlowRule` in the hencky constitutive law the restart does not work. In the case of using a linear elastic law the whole initialize can be skipped, but for the hencky law it didn't work.

philbucher replied:
I would pass the ProcessInfo.

Then it is consistent with other parts of the code and everyone can do what they need

To me this looks like a leftover

rubenzorrilla replied:
Aside of the `ProcessInfo` discussion. This seems a bug in the serialization of one of the constitutive laws...

andinoboerst replied:
True. I did not verify the serialization within the constitutive laws, but I will check if the `mpMPMFlowRule` is skipped and in that case if I can add it.

andinoboerst replied:
This seems to be the problem here:

https://github.com/KratosMultiphysics/Kratos/blob/057edcf143648d0ece0df3595d50bd8f0c44063a/applications/ParticleMechanicsApplication/custom_constitutive/hardening_laws/particle_hardening_law.hpp#L259-L273

andinoboerst replied:
I cannot come up with a solution on how to solve the serializer problem for the particle hardening law (I don't know if someone else had this problem with the recursive call before?). To be fair the information saved there only includes the material properties from the `Materials.json` file, so even when restarting these values can be imported again.
However, this means it needs to be initialized, which goes back to the initial problem of this strange workaround. To solve this workaround it's necessary to either pass the `rProcessInfo` to the `InitializeMaterial()` function or to include a new function for example `InitializeMaterialRestart()` in the `constitutive_law.h`. Maybe the second option does not make so much sense if this application is the only one that needs it, even though in my case it would be better since I would not need to include an if-statement in cases where the whole initialization can be skipped. Both of these solutions include modifying the core though.

andinoboerst replied:
@philbucher @rubenzorrilla @AlejandroCornejo If you all agree I would open an issue assigning the technical committee to decide what is the best way to proceed. I would suggest just passing the `ProcessInfo` as it will be the easiest way to implement this and could be useful in the future as well.

matekelemen replied:
> This seems to be the problem here:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/057edcf143648d0ece0df3595d50bd8f0c44063a/applications/ParticleMechanicsApplication/custom_constitutive/hardening_laws/particle_hardening_law.hpp#L259-L273

**Ugh**, if that can't be solved, I'd at least do one of the following:
- throw an exception when `save` is called (best indicator of failure in my opinion: fails exactly when the problem arises)
- throw an exception when `load` is called (not great, not terrible: fails when your results are about to get messed up)
- issue warnings in `save` and `load` (suboptimal: noone cares about warnings until they've hunted the bug for a couple of days)

andinoboerst replied:
The problem with throwing an exception is that the restart application would never work in this case. When the restart application tries to save or load this information it would then fail every time.
As @philbucher mentioned in the next comment, the other applications do not save these properties, but instead, pass the properties whenever necessary from the outside.
However, if this is to be changed, then all calls involving the `particle_hardening_law` would need to be changed.

matekelemen replied:
Well, if there's no way of omitting them during serialization/deserialization, all that's left is at least to issue a warning.

matekelemen replied:
By the way, I think it would be best to follow up on why serialization is not possible. The current behaviour is a bug and no amount of workarounds will change that. Furthermore, it might be straightforward for us right now to follow how the restarted initialization happens, but I guarantee that it won't be for others who take a look at it for the first time.

I haven't looked into it yet, but I have a sneaking suspicion that we'll find other bugs if we start looking into this more closely. Storing `shared_ptr`s to each other definitely smells like a memory leak.

andinoboerst replied:
I agree, I will look into this. I will try to change the current implementation to avoid storing the material properties in this hardening_law class and instead pass the properties to all of the functions that need them. Maybe this way it would also not be necessary to have an `is_restarted` flag in the `InitializeMaterial()` function.

I will follow up on that.

andinoboerst replied:
Ok so I deleted the attribute `mpProperties` inside the `particle_hardening_law` now and instead pass it as a parameter to all the functions that need it. This way the `mpProperties` do not need to be serialized anymore and thus avoid storing pointers to each other.
See PR #10667.

andinoboerst committed:
Pass ProcessInfo to InitializeMaterial()

andinoboerst committed:
Updated plastic flow serialization

andinoboerst committed:
Newtonian uid was added which needs to be modified for restart
Merge remote-tracking branch 'origin/master' into mpm/restart_application

Newtonian fluid was added which needs to be modified for restart

andinoboerst committed:
Update newtonian fluid law

philbucher commented on kratos/sources/constitutive_law.cpp:
```diff
@@ -548,13 +548,21 @@ bool ConstitutiveLaw::IsIncremental()
 {
 }
 
+void ConstitutiveLaw::InitializeMaterial(const Properties& rMaterialProperties,
```
LGTM!

Please add a deprecation warning to the other function (KRATOS_DEPRECATED_MESSAGE)

philbucher commented on kratos/python/add_constitutive_law_to_python.cpp:
```diff
@@ -209,7 +228,8 @@ void  AddConstitutiveLawToPython(pybind11::module& m)
     .def("FinalizeMaterialResponseCauchy",&ConstitutiveLaw::FinalizeMaterialResponseCauchy)
-    .def("InitializeMaterial",&ConstitutiveLaw::InitializeMaterial)
+    .def("InitializeMaterial",&ConstitutiveLawInitializeMaterialStandard, py::return_value_policy::reference_internal)
+    .def("InitializeMaterial",&ConstitutiveLawInitializeMaterialProcessInfo, py::return_value_policy::reference_internal)
```
Please use py::overload_cast (and remove the return value stuff since it is a void fct)

andinoboerst replied:
Yes thanks, done!

philbucher reviewed (commented):
Nice!

Btw regarding the original problem with the serialzation, I think that CLaws should not save Properties, but use them as passed from outside

andinoboerst committed:
Small fixes
